src/H5Fvfd_swmr.c

```diff
@@ -0,0 +1,2568 @@
+
+    if (closing) { /* For file close */
+
+        /* Close the md file */
+        if (HDclose(shared->vfd_swmr_md_fd) < 0)
+            HSYS_GOTO_ERROR(H5E_FILE, H5E_CANTCLOSEFILE, FAIL, "unable to close the metadata file")
+        shared->vfd_swmr_md_fd = -1;
+
+        /* For VFD SWMR testing: invoke callback if set to generate metadata file checksum */
+        if (shared->generate_md_ck_cb) {
+            if (shared->generate_md_ck_cb(shared->md_file_path_name, shared->updater_seq_num) < 0)
+                HGOTO_ERROR(H5E_FILE, H5E_SYSTEM, FAIL, "error from generate_md_ck_cb()")
+        }
+
+        /* Unlink the md file */
+        if (HDunlink(shared->md_file_path_name) < 0)
+            HSYS_GOTO_ERROR(H5E_FILE, H5E_CANTREMOVE, FAIL, "unable to unlink the metadata file")
+
+        shared->md_file_path_name = (char *)H5MM_xfree(shared->md_file_path_name);
+
+        /* Close the free-space manager for the metadata file */
+        if (H5MV_close(f) < 0)
+            HGOTO_ERROR(H5E_FILE, H5E_CANTRELEASE, FAIL,
+                        "unable to close the free-space manager for the metadata file")
+
+        /* Free the delayed list */
+        while ((curr = TAILQ_FIRST(&shared->shadow_defrees)) != NULL) {
+            TAILQ_REMOVE(&shared->shadow_defrees, curr, link);
+            H5FL_FREE(shadow_defree_t, curr);
+        }
+
+        HDassert(TAILQ_EMPTY(&shared->shadow_defrees));
+
+        if (shared->vfd_swmr_config.generate_updater_files) {
+            if (H5F__generate_updater_file(f, 0, FINAL_UPDATE_FLAG, md_hdr_image, H5FD_MD_HEADER_SIZE,
+                                           md_idx_image, shared->writer_index_offset,
+                                           H5FD_MD_INDEX_SIZE(0)) < 0)
+                HGOTO_ERROR(H5E_FILE, H5E_SYSTEM, FAIL, "can't generate updater file")
+        }
+    }
+    else { /* For file flush */
+        /* Update end_of_tick */
+        if (H5F__vfd_swmr_update_end_of_tick_and_tick_num(shared, TRUE) < 0)
+            HDONE_ERROR(H5E_FILE, H5E_CANTSET, FAIL, "unable to update end of tick")
+    }
+#if 1 /* Save the end of close info. to the log file, subject to comment out. */
+    if (closing)
+        H5F_POST_VFD_SWMR_LOG_ENTRY(f, FILE_CLOSE, "VFD SWMR File close ends");
+#endif
+done:
+
+    /* Stop the timer and close the VFD SWMR log file if it is turned on.
+     * TODO: Please REVIEW to ensure this is the right place to
+     * close the log file.
+     */
+    if (shared->vfd_swmr_log_on && closing) {
+        H5_timer_stop(&(shared->vfd_swmr_log_start_time));
+        HDfclose(shared->vfd_swmr_log_file_ptr);
+    }
+
+    FUNC_LEAVE_NOAPI(ret_value)
+}
+
+/*-------------------------------------------------------------------------
+ * Function: H5F__shadow_image_defer_free
+ *
+ * Purpose:
+ *
+ * Return:   SUCCEED/FAIL
+ *-------------------------------------------------------------------------
+ */
+static herr_t
```

Comment:
All the static routines in this file need FUNC_ENTER/LEAVE macros and proper error handling.